src/Microsoft.DotNet.Wpf/src/PresentationCore/System/Windows/Duration.cs

```diff
@@ -512,7 +512,7 @@ public override string ToString()
         {
             if (HasTimeSpan)
             {
-                return TypeDescriptor.GetConverter(_timeSpan).ConvertToString(_timeSpan);
+                return _timeSpan.ToString();
```

Comment:
This obviously breaks any custom descriptors for `TimeSpan`.

```C#
class HahaConverter : TypeConverter
{
    public override object ConvertTo(ITypeDescriptorContext context, CultureInfo culture, object value, Type destinationType) => "haha";
}

static void Main(string[] args)
{
    TypeDescriptor.AddAttributes(typeof(TimeSpan), new TypeConverterAttribute(typeof(HahaConverter)));
    string s = new Duration(default).ToString(); // returns haha before PR but not after PR
}
```

and look:

https://github.com/thomhurst/ReadableTimeSpan/blob/7eb0c92ebf4132010dbe46b84bf4db8d7dad5b58/TomLonghurst.ReadableTimeSpan/ReadableTimeSpan.cs#L32

https://github.com/Splamy/Sybtitle/blob/ba306e0228c6215246a312ff14dd7e266bdeec49/Sybtitle/HumanTimeSpan.cs#L185-L192

Reply:
It seemed like an oversight to me, I was checking other related `Timeline` types like `RepeatBehavior`, and there it calls regular `ToString` on `TimeSpan`. But it is a behavioral change I didn't consider, indeed.

In my own opinion, I wouldn't want the behavior of `Duration`'s `ToString` change based on a `TimeSpan`'s current converter set via attributes; since `ConvertFrom` won't change accordingly.

My bigger problem is that we're doing always conversion on current culture tbh, there's so much inconsistencies it's killing me with the converters.

Whether it's exceptions throwing, sometimes you throw in `CanConvert`* methods because of bugs (e.g. casts to types).

Reply:
Regarding my culture problem: I think I remember why, `TimeSpan` got globalized later on afaik, maybe `net4`?

So the support was then never added to the converters most likely.

Reply:
Yeah obviously the `ConvertFrom` and `ConvertTo` should be consistent. I would think that is a bug in the converter more than anything else. It's reasonable for people to just use `Duration` for non-negative timestamps in general, even in non-WPF applications relying on `ToString()`.

I am not that confident on it being an oversight, you need extra effort and knowledge to write this. Maybe it was needed for the designer?

Unlike numbers and dates, the `TimeSpan` is not culturally sensitive unless you explicitly ask for it, it uses the constant invariant format.

I agree that a better consistency would be desirable across converters.

Reply:
Thank you by the way, I love your reviews, you catch some great things everytime.

In your opinion then, we would revert the change on `Duration`'s `ToString` and handle it independently in the Converter? I think one internal method can solve this without breaking anything. We can maybe follow-up the `ToString` in another issue.

I was going through the WPF tests and none of them regarding `Duration` counts on this.

Yeah, I just went through the docs to refresh my memory on `TimeSpan` and was correct indeed, the culture-sensitivity didn't arrive until `net4`.

Based on that, I went to check the old .NET 3.5 libraries, and yeah, the culture overload gets added in `net4` and subsequently used in `TimeSpanConverter`'s `ConvertFrom` but obviously all the WPF code is left untouched, indirectly swallowing this change because it was passing the `cultureInfo` there. So that's how `ConvertFrom` became sensitive to culture passed in but `ConvertTo` is always culture-insensitive. So yeah this feels like a bug.

Yeah I was hoping later on to summarize the differences in one issue between the converters but first I have to get through them, write tests for them to catch all edge cases and get some output out of this, then maybe some action.

Reply:
It wasn't a breaking change, right? Without specifying a format, converting to string still uses the invariant format regardless of culture, and parsing still supports the invariant one but also became tolerant to a localized format.

Yeah I would think reverse the `Duration.ToString()` for now, especially since you are selling this as an optimization in `DurationConverter`, and we should probably use `TimeSpanConverter` in the converter both ways for consistency. That would also save you the type descriptor lookup on every call.

Reply:
Well partially it is, because let's say you've been passing to `DurationConverter`/`TimeSpanConverter` a string with, let's say "07:12:14:45.3448" (that could have been generated via `ToString()`) and `CurrentCulture` (or `null`, that's more likely), the tokenizer in `TimeSpan` will call https://learn.microsoft.com/en-us/dotnet/api/system.globalization.datetimeformatinfo.getinstance?view=net-8.0 under the hood.

Pre-`net4` this would work, as it was ignored obviously since you weren't passing the `CultureInfo` beyond `TimeSpanConverter`, but when it swapped from `Parse(string s)` overload in `net4` to `Parse(string input, IFormatProvider? formatProvider)` and the culture is recognized, this blows on systems with different decimal separator than `.` for example.

Essentially `ConvertTo`/`ConvertFrom` gives you incompatible values (throws) since then if you need cultural sensitivity or you were counting on it pre-`net4`. Sure there was a compat switch but I'm not sure whether it covered the converter case.

Reply:
Yes, because you're using `var`. Just kidding.

```csharp
//This is the original behaviour pre-net 4.0 of TimeSpan converter, since it was the only existent overload
//Therefore passing any CultureInfo had no effect
CultureInfo czechCulture = new CultureInfo("cs-CZ");
TimeSpan timeStamp =  TimeSpan.Parse("07:12:14:45.3448");
string xx =  timeStamp.ToString();

//This is post net4 behaviour, when the overload in TimeSpanConverter was changed to this one
//So any null passed in to the converter, or specific CultureInfo with different separators etc. 
//ends up with a beautiful exception
Thread.CurrentThread.CurrentCulture = czechCulture;
timeStamp = TimeSpan.Parse("07:12:14:45.3448", null);
xx = timeStamp.ToString();
```

Reply:
You might be getting exception because your string is invalid, cs-CZ uses decimal comma... 

```C#
Thread.CurrentThread.CurrentCulture = CultureInfo.GetCultureInfo("cs-CZ");
string sInv = "07.12:14:45.3448";
string sLoc = "07:12:14:45,3448";
var t1 = TimeSpan.Parse(sInv, null);
var t2 = TimeSpan.Parse(sLoc, null);
var var = t1.ToString(); // 7.12:14:45.3448000
```

doesn't throw either...?

Reply:
My string will parse as `InvariantCulture` passed in for example; but I'm not sure whether you still get my point.

`DurationConverter`/`TimeSpanConverter` pre-net4 uses (there's nothing else to use but you're still passing `CultureInfo` to the converter, e.g. `null`): `TimeSpan.Parse("07:12:14:45.3448");`

`DurationConverter`/`TimeSpanConverter` post-net4 uses the new overload (now your `CultureInfo` starts to matter, even if it was null): `TimeSpan.Parse("07:12:14:45.3448", cultureInfo);`

And that's a behavioral change, because now the string started to throw on different cultures instead of being converted.

Reply:
Well either we are talking each about something else or one of us is wrong. I am claiming that if your string is a valid `TimeSpan` in invariant/pre-net4 format, then both `TimeSpan.Parse(s)` and `TimeSpan.Parse(s, cultureInfo)` behave the same for any culture provided, i.e. they always succeed, no new exceptions are thrown. The parser always tries the invariant/pre-net4 format, as I linked in the code. What changed is that now strings that were not valid before will get parsed if they match the format in the culture supplied.

I am also happy to cease the discussion because it doesn't matter very much here. :)

Reply:
Yeah let's not waste anymore time on this haha; anyways here you go and I hope you will understand now the edge-case.

[Sharplab demo](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AcQBsJgBDdgSwBe3DHwh4A3NTqMWAFQAWsbgBM+eAOaSaAZkYl6AYXoBvavXMykjFPQCy3NQAoAlCbMWPAZQxQ16+vwYMFC89AC89ABEtADsIGQk8SggKACszNooKAAckVoeHu4F9AD0JQDqMADksPTcALbBfGDceADORQUKSsrMBjhQsHgY3TAqfQNDGP3sGAMw4fR4MADuhjiz8wCSeABmEI6RMHhwAKqekc75xfSdHmVMMY1Q9MoQMG30GPJ8HVQ3PgAnsZ6AAFWBBPDKLb1AxiABuwQwADEoBB6o5AsFeGglqt1ptYDt9ocwG04AYAFqXXG7XhtGBXegAXzuFhaGDA8nojmR0HqwgAoggwDAAA4iMSuYxMACch2WiJe8m4YrFxzal3ErP+BSBJjBEOO0NhCKRqPRmL4QRC7FxyzWMzmRL2B0iZIp1OcuJ8OEZ4hZbPMHK5PL5UAFGGFoolojw0rlhxWMHYkEaXwgeIwKC1Oo8eYsnSYVlkfEanjFrUNMEhJrheCVKLRGKYDBE5creG8vg0uKd21d9DAG2dMGJEFxwAgEHY9BwDIAcjWUM5OqZdcU+LsefOYEvs6uNwV1zcCkQYvRSx3WsxQdwoAzHO2YBXWt2/Ljh4Sx66mcUg4GR4eCmDIASep4eOel5li+na3vej7Pq+XY+H4VwAQW5g6phQA===)

[SharpLab demo 2](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoIgBgAIiyA6AcQBsJgBDdgSwBe3DHwh4A3NTqMWAFQAWsbgBM+eAOaSaAZkYl6AYXoBvavXMykjFPQCy3NQAoAlCbMWPAZQxQ16+vwYMFC89AC89ABEtADsIGQk8SggKACsaNooKAAckVoeHu4F9AD0JQDqMADksPTcALbBfGDceADORQUKSsrMBjhQsHgY3TAqfQNDGP3sGAMw4fR4MADuhjiz8wCSeABmEI6RYG1wBgBakc75xfSdNwVlTDGNUPTKEDBt9BjyfB1UNx8AE9jPQAAqwIJ4ZRbeoGMQAN2CGAAYlAIPVHIFgrw0OtNrAdvtmDsEdxfK1phs5rA8bteG0YFd6ABfO4WFoYMDyeiOFHQerCACiCDAMAADiIxK5jEwAJyHZZI17ybji8UwdqXcRsgEFYEmcGQzUwuGI5FojFYvhBELsPEzGkwIkQEl4MkU4aO+Z4nw4JniVns8yc7m8/lQQUYEViyWiPAy+WHFYwdiQRrfCBLGAYFDa3UeAsWTpMKyyPiNTzi1pGnMm2Hw90W9GYpgMESV6t4by+DQO6nbPZZsADwlDvHACAQdj0HCMgByOZQzk6pj1xT4u15c5gi9zK/XBTX9w8RBi9HLndazDB5MZjg7MCrrR7fjxI4JzqHzOKwaDh48VNGT/Y8TwsM8LwrJ8uxvO8YAfKDn27Hw/CuP8i3MXUMKAA)